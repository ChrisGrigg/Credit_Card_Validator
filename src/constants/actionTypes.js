export const VALIDATE_CREDIT_CARD = 'VALIDATE_CREDIT_CARD';
