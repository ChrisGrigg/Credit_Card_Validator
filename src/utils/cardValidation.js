export function getCardType(settings) {
  if (!settings.cardNumber) {
    throw new Error('Card number property does not exist');
  }

  let number = settings.cardNumber;
  number = ('' + number).replace(/\D/g, '');

  // TODO: with more time I would implement these checks as strategies,
  // TODO: compose them into this web component and call a validate() method on each
  // TODO: this would make it easily extensible and make it straightforward to create mocks and inject rules in acceptance tests etc
  if (/^5[1-5]/.test(number) && number.length === 16) {
    return 'MasterCard';
  } else if (/^(6011|622|64|65)/.test(number) && number.length === 16) {
    return 'Discover';
  } else if (/^3[47]/.test(number) && number.length === 15) {
    return 'AMEX';
  } else if (/^(4)/.test(number) && (number.length === 13 || number.length === 16)) {
    return 'VISA';
  }

  return 'Unknown';
}

export function executeLuhnAlgorithm(settings) {
  if (!settings.cardNumber) {
    throw new Error('Card number property does not exist');
  }

  // https://gist.github.com/DiegoSalazar/4075533
  if (/[^0-9-\s]+/.test(settings.cardNumber)) return false;

  let nCheck = 0;
  let nDigit = 0;
  let bEven = false;
  for (let n = settings.cardNumber.length - 1; n >= 0; n--) {
    let cDigit = settings.cardNumber.charAt(n);
    nDigit = parseInt(cDigit, 10);

    if (bEven) {
      if ((nDigit *= 2) > 9) nDigit -= 9;
    }

    nCheck += nDigit;
    bEven = !bEven;
  }

  return (nCheck % 10) === 0;
}

export function requiredDataProvidedForValidation(settings) {
  return settings.cardNumber > 0;
}

export function validateCardNumber(settings) {
  const cardType = getCardType(settings);
  const validCardType = cardType !== 'Unknown' && executeLuhnAlgorithm(settings);
  return {
    cardNumber: settings.cardNumber,
    cardType: cardType,
    valid: validCardType
  }
}
