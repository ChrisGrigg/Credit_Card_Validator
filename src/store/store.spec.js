import * as ActionTypes from '../constants/actionTypes';

import configureStore from './configureStore';

import {validateCardNumber} from '../utils/cardValidation';

describe('Store', () => {
  it('should display results when necessary data is provided', () => {
    const store = configureStore();

    const actions = [
      { type: ActionTypes.VALIDATE_CREDIT_CARD, settings: store.getState(), fieldName: 'cardNumber', value: '4111111111111111' },
    ];
    actions.forEach(action => store.dispatch(action));

    const actual = store.getState();
    const expected = {
      cardNumber: '4111111111111111',
      requiredDataProvidedForValidation: true,
      displayResults: false,
      cardNumberValid: validateCardNumber(store.getState().cardValidation)
    };

    expect(actual.cardValidation).toEqual(expected);
  });

  it('should not display results when necessary data is not provided', () => {
    const store = configureStore();

    const actions = [
      // { type: ActionTypes.VALIDATE_CREDIT_CARD, settings: store.getState(), fieldName: 'cardNumber', value: '4111111111111111' },
    ];

    actions.forEach(action => store.dispatch(action));

    const actual = store.getState();

    const expected = {
      cardNumber: '',
      displayResults: false,
      requiredDataProvidedForValidation: false,
      cardNumberValid: {
        cardNumber: '',
        cardType: '',
        valid: false
      }
    };

    expect(actual.cardValidation).toEqual(expected);
  });


  it('should handle a flurry of actions', () => {
    const store = configureStore();

    const actions = [
      { type: ActionTypes.VALIDATE_CREDIT_CARD, settings: store.getState(), fieldName: 'cardNumber', value: '4111111111111111' },
      { type: ActionTypes.VALIDATE_CREDIT_CARD, settings: store.getState(), fieldName: 'cardNumber', value: '4111111111111' },
      { type: ActionTypes.VALIDATE_CREDIT_CARD, settings: store.getState(), fieldName: 'cardNumber', value: '4012888888881881' },
      { type: ActionTypes.VALIDATE_CREDIT_CARD, settings: store.getState(), fieldName: 'cardNumber', value: '378282246310005' },
    ];
    actions.forEach(action => store.dispatch(action));

    validateCardNumber(store.getState().cardValidation);

    const moreActions = [
      { type: ActionTypes.VALIDATE_CREDIT_CARD, settings: store.getState(), fieldName: 'cardNumber', value: '6011111111111117' },
      { type: ActionTypes.VALIDATE_CREDIT_CARD, settings: store.getState(), fieldName: 'cardNumber', value: '5105105105105100' },
      { type: ActionTypes.VALIDATE_CREDIT_CARD, settings: store.getState(), fieldName: 'cardNumber', value: '5105105105105106' },
      { type: ActionTypes.VALIDATE_CREDIT_CARD, settings: store.getState(), fieldName: 'cardNumber', value: '9111111111111111' },
    ];

    moreActions.forEach(action => store.dispatch(action));

    const actual = store.getState();
    expect(actual.cardValidation).toMatchSnapshot();
  });
});
