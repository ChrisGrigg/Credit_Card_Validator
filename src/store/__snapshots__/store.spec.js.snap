// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store should handle a flurry of actions 1`] = `
Object {
  "cardNumber": "9111111111111111",
  "cardNumberValid": Object {
    "cardNumber": "9111111111111111",
    "cardType": "Unknown",
    "valid": false,
  },
  "displayResults": false,
  "requiredDataProvidedForValidation": true,
}
`;
